src/Stepping.TmProviders.LocalTm/Stepping/TmProviders/LocalTm/TransactionManagers/LocalTmManager.cs

```
@@ -0,0 +1,253 @@
+            return false;
+        }
+
+        var tmTransactionModel = await GetAsync(gid, cancellationToken);
+
+        if (await IsExpectedStatusAsync(tmTransactionModel, LocalTmConst.StatusPrepare))
+        {
+            return false;
+        }
+
+        try
+        {
+            if (!await QueryPreparedAsync(tmTransactionModel, cancellationToken))
+            {
+                await UpdateRollbackAsync(tmTransactionModel, LocalTmConst.ReasonPrepareFailure, cancellationToken);
+
+                Logger.LogInformation("Local transaction '{gid}' rolled back.", gid);
+
+                return false;
+            }
+
+            await UpdateSubmitAsync(tmTransactionModel, cancellationToken);
+
+            Logger.LogInformation("Local transaction '{gid}' committed.", gid);
+
+            return true;
+        }
+        catch (Exception ex)
+        {
+            await UpdateNextRetryAsync(tmTransactionModel, cancellationToken);
+
+            Logger.LogWarning(ex, "Local transaction '{gid}' process query prepared failed. Wait for the next retry.", gid);
+
+            return false;
+        }
+    }
+
+    protected virtual async Task<bool> QueryPreparedAsync(TmTransactionModel tmTransactionModel, CancellationToken cancellationToken)
+    {
+        var dbContextLookupInfoModel = tmTransactionModel.SteppingDbContextLookupInfo;
+        var dbContextProvider = await DbContextProviderResolver.ResolveAsync(dbContextLookupInfoModel.DbProviderName);
+        var dbContext = await dbContextProvider.GetAsync(dbContextLookupInfoModel);
+
+        var barrierInfoModel = await BarrierInfoModelFactory.CreateForRollbackAsync(tmTransactionModel.Gid);
+
+        var barrierInserter = await DbBarrierInserterResolver.ResolveAsync(dbContextLookupInfoModel.DbProviderName);
+
+        if (await barrierInserter.TryInsertBarrierAsync(barrierInfoModel, dbContext, cancellationToken))
```

Insert and ignore the exception, and then query the barrier again.

Think about this case: the TM crashes before setting the transaction result.

---

I think `TryInsertBarrierAsync` already handles this case.